sel/src/Sel/SecretKey/EncryptedStream.hs

```diff
@@ -0,0 +1,542 @@
+-- Copyright: (C) Hécate Moonlight 2022
+-- License: BSD-3-Clause
+-- Maintainer: The Haskell Cryptography Group
+-- Portability: GHC only
+module Sel.SecretKey.EncryptedStream
+  ( -- ** Secret Key
+    SecretKey
+  , newSecretKey
+
+    -- ** CipherText
+  , CipherText
+  , cipherTextFromByteString
+  , cipherTextToBinary
+  , cipherTextToHexByteString
+  , cipherTextToHexText
+
+    -- ** Stream Header
+  , Header
+  , headerFromByteString
+
+    -- ** Stream Tags
+  , StreamTag (..)
+  , streamTagToConstant
+  , tagConstantToStreamTag
+
+    -- ** Error type
+  , EncryptedStreamError (..)
+
+    -- ** Stream Operations
+  , Multipart (..)
+  , encryptStream
+  -- , decryptStream
+
+    -- *** Encryption
+  , initPushStream
+  , pushToStream
+
+    -- *** Decryption
+  , StreamResult (..)
+  , initPullStream
+  , pullFromStream
+
+    -- *** Key regeneration
+  , rekey
+  ) where
+
+import Control.Monad.IO.Class (MonadIO, liftIO)
+import Data.ByteString (StrictByteString)
+import qualified Data.ByteString as BS
+import qualified Data.ByteString.Base16 as Base16
+import qualified Data.ByteString.Internal as BS
+import qualified Data.ByteString.Unsafe as BS
+import Data.Kind (Type)
+import Data.Text.Display (Display (..), OpaqueInstance (..), ShowInstance (..))
+import Foreign (ForeignPtr, Ptr, Word8)
+import qualified Foreign
+import Foreign.C (CChar, CSize, CUChar, CULLong)
+import GHC.IO.Handle.Text (memcpy)
+import System.IO.Unsafe (unsafeDupablePerformIO)
+
+import Data.Text (Text)
+import LibSodium.Bindings.SecretStream
+  ( CryptoSecretStreamXChaCha20Poly1305State
+  , cryptoSecretStreamXChaCha20Poly1305ABytes
+  , cryptoSecretStreamXChaCha20Poly1305HeaderBytes
+  , cryptoSecretStreamXChaCha20Poly1305InitPull
+  , cryptoSecretStreamXChaCha20Poly1305InitPush
+  , cryptoSecretStreamXChaCha20Poly1305KeyBytes
+  , cryptoSecretStreamXChaCha20Poly1305KeyGen
+  , cryptoSecretStreamXChaCha20Poly1305Pull
+  , cryptoSecretStreamXChaCha20Poly1305Push
+  , cryptoSecretStreamXChaCha20Poly1305Rekey
+  , cryptoSecretStreamXChaCha20Poly1305StateBytes
+  , cryptoSecretStreamXChaCha20Poly1305TagFinal
+  , cryptoSecretStreamXChaCha20Poly1305TagMessage
+  , cryptoSecretStreamXChaCha20Poly1305TagPush
+  , cryptoSecretStreamXChaCha20Poly1305TagRekey
+  )
+import Sel.Internal
+
+-- | The 'SecretKey' is used to encrypt the stream.
+--
+-- @since 0.0.1.0
+newtype SecretKey = SecretKey (ForeignPtr CUChar)
+  deriving
+    ( Display
+      -- ^ @since 0.0.1.0
+      -- > display secretKey == "[REDACTED]"
+    )
+    via (OpaqueInstance "[REDACTED]" SecretKey)
+
+-- |
+--
+-- @since 0.0.1.0
+instance Eq SecretKey where
+  (SecretKey hk1) == (SecretKey hk2) =
+    unsafeDupablePerformIO $
+      foreignPtrEq hk1 hk2 cryptoSecretStreamXChaCha20Poly1305KeyBytes
+
+-- |
+--
+-- @since 0.0.1.0
+instance Ord SecretKey where
+  compare (SecretKey hk1) (SecretKey hk2) =
+    unsafeDupablePerformIO $
+      foreignPtrOrd hk1 hk2 cryptoSecretStreamXChaCha20Poly1305KeyBytes
```

Comment:
We do this in a bunch of places. Might be worth writing a helper `newtype` for DerivingVia for these.